File: LAB3/sjf.js
// class job{

//     constructor(burst,arrival){
//     var burst,arrival
//     this.burst=burst
//     this.arrival=arrival
//     }
//     method_getburst(){
//         return this.burst
//     }
//     method_getarrival(){
//         return this.arrival
//     }
// }

// let job_list=[new job(6,2),new job(2,5),new job(8,1),new job(3,0),new job(4,4)]

let job_list = [
        [1, 3, 2],
        [2, 4, 0],
        [3, 2, 4],
        [4, 4, 5],
    ] //id,burst time,arrival time
    // function swap(a,b){
    //     let temp=a
    //     a=b
    //     b=temp
    // }
function arrange_arrival(number, list) {
    for (let i = 0; i < number; i++) {
        for (let j = 0; j < number - i - 1; j++) {
            if (list[j][2] > list[j + 1][2]) {
                let temp = list[j]
                list[j] = list[j + 1]
                list[j + 1] = temp
            }
        }
    }
    return list
}

function complete_time(number, list) {
    var temp, value
    list[0].push(list[0][2] + list[0][1])
    list[0].push(list[0][3] - list[0][2])
    list[0].push(list[0][4] - list[0][1])
    console.log(list[0])
    for (let i = 1; i < number; i++) {
        temp = list[i - 1][3]
        let low = list[i][1]
        for (let j = i; j < number; j++) {
            if (temp >= list[j][2] && low >= list[j][1]) {
                low = list[j][1]
                value = j
            }
        }
        list[value].push(temp + list[value][1])
        list[value].push(list[value][3] - list[value][2])
        list[value].push(list[value][4] - list[value][1])

        temp = list[value]
        list[value] = list[i]
        list[i] = temp
    }
    return list
}
job_list = arrange_arrival(job_list.length, job_list)
console.log(job_list)
let sjf = complete_time(job_list.length, job_list)

console.log("ID\t Burst time\t Arrival time\t waiting time\t turnaround time")
for (let i = 0; i < sjf.length; i++) {
    console.log(sjf[i][0] + "\t " + sjf[i][1] + "\t\t " + sjf[i][2] + "\t\t " + sjf[i][5] + "\t\t " + sjf[i][4])
}

File: LAB3/SJF.py
'''
Shortest Job First
Your program receives as input the number of jobs waiting in queue
and the time required to execute each job. 
Display the outcome of each scheduling algorithm.
INDEXES
0 - Job number
1 - Burst time
2 - Arrival time
3 - Completion Time
4 - Waiting time = Turnaround time - Burst time
5 - Turnaround time = Completion time - Arrival time 
'''


def print_times(job_list):
    print("JN      BT      AT      CT      WT      TAT")
    for i in range(len(job_list[0])):
        for j in range(0, 6):
            print(job_list[j][i], end="\t")
        print()
    print()


def swap(job_list, i, j):
    for idx in range(0, 6):
        job_list[idx][i], job_list[idx][j] = job_list[idx][j], job_list[idx][i]


def sort_arrival_times(job_list):
    for i in range(len(job_list[0])):
        for j in range(len(job_list[0])):
            if i == j:
                continue
            elif job_list[2][i] < job_list[2][j]:
                swap(job_list, i, j)
    print("Ordered According to Arrival Times")
    print_times(job_list)


def calculate_times(job_list):
    temp_job = -1
    # Calculate for first job
    # Completion Time = Arrival Time + Burst Time
    job_list[3][0] = job_list[2][0] + job_list[1][0]
    # Turnaround Time = Completion Time - Arrival Time
    job_list[5][0] = job_list[3][0] - job_list[2][0]
    # Waiting Time = Turnaround Time - Burst Time
    job_list[4][0] = job_list[5][0] - job_list[1][0]

    for i in range(1, len(job_list[0])):
        previous_completion_time = job_list[3][i-1]
        current_shortest_burst = job_list[1][i]

        for j in range(i, len(job_list[0])):
            if previous_completion_time >= job_list[2][j] and job_list[1][j] <= current_shortest_burst:
                current_shortest_burst = job_list[1][j]
                temp_job = j

        # Completion Time = Arrival Time + Burst Time
        job_list[3][temp_job] = previous_completion_time + job_list[1][temp_job]
        # Turnaround Time = Completion Time - Arrival Time
        job_list[5][temp_job] = job_list[3][temp_job] - job_list[2][temp_job]
        # Waiting Time = Turnaround Time - Burst Time
        job_list[4][temp_job] = job_list[5][temp_job] - job_list[1][temp_job]

        # Swap in case the job has arrived and has a shorter burst time
        swap(job_list, temp_job, i)


def shortest_job_first(job_list):
    sort_arrival_times(job_list)
    calculate_times(job_list)



job_number = 5
job_list = [[0 for j in range(job_number)] #Create array of [[],[],[],[]]
            for i in range(6)]                 
job_list[0] = [i for i in range(1, job_number + 1)] #job number

job_list[1] = [6, 2, 8, 3, 4] #Burst time

job_list[2] = [2, 5, 1, 0, 4] #Arrival time
print("Original Input")
print_times(job_list)

shortest_job_first(job_list)

print("Final Result")
print_times(job_list)
print("Average Turnaround Time:",sum(job_list[5])/job_number,end="s\n")
print("Average Waiting Time:",sum(job_list[4])/job_number,end="s")